pkg/kube/client.go

```diff
@@ -373,12 +401,31 @@ func (c *Client) Update(namespace string, originalReader, targetReader io.Reader
 			c.Log("Failed to delete %q, err: %s", info.Name, err)
 		}
 	}
-	if shouldWait {
-		return c.waitForResources(time.Duration(timeout)*time.Second, target)
+	if opts.ShouldWait {
+		err := c.waitForResources(time.Duration(opts.Timeout)*time.Second, target)
+
+		if opts.CleanupOnFail && err != nil {
+			cleanupErrors = c.cleanup(newlyCreatedResources)
+			return fmt.Errorf(strings.Join(append([]string{err.Error()}, cleanupErrors...), " && "))
+		}
+
+		return err
 	}
 	return nil
 }
 
+func (c *Client) cleanup(newlyCreatedResources []*resource.Info) (cleanupErrors []string) {
+	for _, info := range newlyCreatedResources {
+		kind := info.Mapping.GroupVersionKind.Kind
+		c.Log("Deleting newly created %s with the name %q in %s...", kind, info.Name, info.Namespace)
```

Comment:
Nit: This log entry could lead to some questions for someone operating Tiller because it doesn't have something like "Cleaning up created resources due to upgrade failure" before it starts running, they'll just have to assume from the error while creating the resource. It might be worth putting some log lines before `cleanup` is called

Reply:
Added some messages outside of `cleanup`:

https://github.com/helm/helm/pull/4871/files#diff-a483d6c0863082c3df21f4aad513afe2R374
https://github.com/helm/helm/pull/4871/files#diff-a483d6c0863082c3df21f4aad513afe2R409

Reply:
Awesome, looks great!